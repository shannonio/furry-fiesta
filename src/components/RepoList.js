import React, { useEffect } from 'react';
import { withRouter } from 'react-router-dom';
import { find } from 'lodash';

import './RepoList.scss';

const RepoList = ({
    repos,
    currentRepo,
    updateCurentRepo,
    fetchRepos,
    match,
    history
  }) => {

  useEffect(() => {
    if (!repos) { fetchRepos() }
  }, []);

  useEffect(() => {
    const repo = find(repos, o => {
      return o.name === match.params.name
    });
    updateCurentRepo(repo);
  }, [repos, match.params.name])

  const isActiveRepoClass = repo => {
    return repo.name === currentRepo.name ? 'RepoList__repos-row--active' : '';
  };

  const changeRepo = repo => history.push(`/repo/${repo.owner.login}/${repo.name}`);

  const repoList = () => repos.map(repo =>(
    <div className={`RepoList__repos-row ${isActiveRepoClass(repo)}`}
         onClick={() => changeRepo(repo)}
         key={repo.id}>
      <span className="RepoList__repos-row--name">
        { repo.name }
      </span>
    </div>
  ));

  return (
    <div className="RepoList">
      <div className="RepoList__repos">
        { repos && repoList() }
      </div>
    </div>
  );
}


export default withRouter(RepoList);
