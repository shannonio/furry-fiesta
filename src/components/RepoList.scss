@import '../styles/core';

.RepoList {
  display: none;

  @media (min-width: 768px) {
    display: block;
  }

  &__repos {
    display: grid;
    float: left;
    grid-template-rows: 1fr;
    padding: 0 1rem;
    align-items: start;
  }

  &__repos-row {
    display: grid;
    grid-template-columns: 1fr;
    justify-items: start;
    padding: 1rem;
    background-color: $light-grey;
    color: $white;

    &:hover, &--active {
      font-weight: bold;
      cursor: pointer;
      background-color: $light-purple;
      transition: 0.218s background-color ease-in-out;
    }
  }
}
