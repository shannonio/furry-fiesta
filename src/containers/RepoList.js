import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import fetchIssues from '../actions/issues';
import fetchRepos from '../actions/repos';
import { updatePrioritizedIssues } from '../actions/prioritizedIssues';
import { updateCurentRepo } from '../actions/currentRepo';

import { getIssues } from '../reducers/issues';
import { getRepos } from '../reducers/repos';
import { getUser } from '../reducers/user';
import { getPrioritizedIssues } from '../reducers/prioritizedIssues';
import { getCurrentRepo } from '../reducers/currentRepo';

import RepoList from '../components/RepoList'

const mapStateToProps = (state, ownProps) => {
  return {
    repos: getRepos(state),
    user: getUser(state),
    currentRepo: getCurrentRepo(state),
  }
};

const mapDispatchToProps = (dispatch) => {
  return {
    fetchRepos: () => fetchRepos()(dispatch),
    updateCurentRepo: repo => dispatch(updateCurentRepo(repo))
  }
};

const RepoListContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(RepoList));

export default RepoListContainer
