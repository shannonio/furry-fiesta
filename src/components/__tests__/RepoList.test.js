import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import RepoList from '../RepoList';

describe('RepoList', () => {
  const repos = [
    { name: 'Snarf', owner: { login: 'Blaster' }, id: '1' },
    { name: 'Mermaid', owner: { login: 'Little' }, id: '2' }
  ];
  const issues = [];

  const initialProps = {
    repos,
    issues,
    match: { params: {
      name: repos[0].name,
      owner: {
        id: repos[0].owner.login
      }
    } },
    updateCurentRepo: jest.fn(),
    history: { push: jest.fn() },
    currentRepo: repos[0],
    fetchRepos: jest.fn(),
    fetchIssues: jest.fn()
  };

  const createComponent = (props = initialProps) => (
    mount(
      <RepoList {...props} />
    )
  )

  it('renders without crashing', () => {
    const component = createComponent();

    expect(component).toMatchSnapshot();
  });

  it('has the correct active repo highlighted', () => {
    const component = createComponent();
    expect(component.find('.RepoList__repos-row--active').text()).toBe(initialProps.currentRepo.name);
  });

  it('changes the currentRepo onClick', () => {
    const component = createComponent();
    const lastRepoIdx = repos.length - 1;
    component.find('.RepoList__repos-row').last().simulate('click');

    expect(initialProps.history.push).toHaveBeenCalledWith(`/repo/${repos[lastRepoIdx].owner.login}/${repos[lastRepoIdx].name}`);
  });

  it('calls updateCurentRepo on url change', () => {
    const component = createComponent();
    const lastRepoIdx = repos.length - 1;

    const props = { ...initialProps };
    props.match = {
      params: {
        name: repos[lastRepoIdx].name,
        owner: {
          id: repos[lastRepoIdx].owner.login
        }
      }
    };

    component.setProps(props);

    expect(initialProps.updateCurentRepo).toHaveBeenCalledWith(repos[lastRepoIdx]);
  });

});
