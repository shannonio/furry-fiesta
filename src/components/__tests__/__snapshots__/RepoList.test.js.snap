// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoList renders without crashing 1`] = `
<RepoList
  currentRepo={
    Object {
      "id": "1",
      "name": "Snarf",
      "owner": Object {
        "login": "Blaster",
      },
    }
  }
  fetchIssues={[MockFunction]}
  fetchRepos={[MockFunction]}
  history={
    Object {
      "push": [MockFunction],
    }
  }
  issues={Array []}
  match={
    Object {
      "params": Object {
        "name": "Snarf",
        "owner": Object {
          "id": "Blaster",
        },
      },
    }
  }
  repos={
    Array [
      Object {
        "id": "1",
        "name": "Snarf",
        "owner": Object {
          "login": "Blaster",
        },
      },
      Object {
        "id": "2",
        "name": "Mermaid",
        "owner": Object {
          "login": "Little",
        },
      },
    ]
  }
  updateCurentRepo={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "id": "1",
            "name": "Snarf",
            "owner": Object {
              "login": "Blaster",
            },
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="RepoList"
  >
    <div
      className="RepoList__repos"
    >
      <div
        className="RepoList__repos-row RepoList__repos-row--active"
        key="1"
        onClick={[Function]}
      >
        <span
          className="RepoList__repos-row--name"
        >
          Snarf
        </span>
      </div>
      <div
        className="RepoList__repos-row "
        key="2"
        onClick={[Function]}
      >
        <span
          className="RepoList__repos-row--name"
        >
          Mermaid
        </span>
      </div>
    </div>
  </div>
</RepoList>
`;
